driver/reducecalculation_driver.hpp

```diff
@@ -237,27 +312,51 @@ int ReduceCalculationDriver<Tgpu, Tref>::AllocateBuffersAndCopy()
     uint32_t ctx = 0;
 
     in_dev        = std::unique_ptr<GPUMem>(new GPUMem(ctx, in_sz, sizeof(Tgpu)));
-    out_dev       = std::unique_ptr<GPUMem>(new GPUMem(ctx, out_sz, sizeof(Tgpu)));
+    out_dev       = IsLogicalCalculation() ? std::make_unique<GPUMem>(ctx, out_sz, sizeof(bool))
+                                           : std::make_unique<GPUMem>(ctx, out_sz, sizeof(Tgpu));
     workspace_dev = std::unique_ptr<GPUMem>(new GPUMem(ctx, ws_sizeInBytes, sizeof(std::byte)));
 
-    in      = std::vector<Tgpu>(in_sz, static_cast<Tgpu>(0));
-    out     = std::vector<Tgpu>(out_sz, static_cast<Tgpu>(0));
-    outhost = std::vector<Tref>(out_sz, static_cast<Tref>(0));
+    in = std::vector<Tgpu>(in_sz, static_cast<Tgpu>(0));
 
-    for(int i = 0; i < in_sz; ++i)
+    if(IsLogicalCalculation())
     {
-        in[i] = prng::gen_A_to_B<Tgpu>(static_cast<Tgpu>(0.0), static_cast<Tgpu>(1.0));
+        logical_out     = std::vector<uint8_t>(out_sz, 0);
+        logical_outhost = std::vector<uint8_t>(out_sz, 0);
+    }
+    else
+    {
+        out     = std::vector<Tgpu>(out_sz, static_cast<Tgpu>(0));
+        outhost = std::vector<Tref>(out_sz, static_cast<Tref>(0));
+    }
+
+    for(size_t i = 0; i < in_sz; ++i)
+    {
+        bool is_zero = prng::gen_A_to_B<int>(0, 2) == 0;
```

Comment:
why you need `is_zero` variable ?

Reply:
My operation is Any/All, which primarily checks for falsity in the input values. For example, if Tgpu is float, I want the input to include some values explicitly set to 0 (indicating False). Without using is_zero, it’s highly unlikely for the random logic to generate 0 values. This setup is solely for testing purposes and does not affect the operation's logic.

Reply:
why don't you use `in[i] = prng::gen_A_to_B<Tgpu>(static_cast<Tgpu>(0.0), static_cast<Tgpu>(2.0));`

Reply:
As I explained earlier, your suggestion works if Tgpu is of an integer type. However, if Tgpu is a floating-point type, the probability of randomly selecting 0 is extremely low with this logic. I want to ensure there are sufficient 0 values in the input to adequately test for False positions

Reply:
it is typo error, should be `in[i] = prng::gen_A_to_B<int>(0, 2);`

Reply:
I understand your point, but using your suggested approach:
```
for (size_t i = 0; i < in_sz; ++i)  
{  
  in[i] = static_cast<Tgpu>(prng::gen_A_to_B<int>(0, 2));  
}
```

This approach does make the code shorter, but using this way, it only generates whole numbers with no decimal places. As a result, the tolerance check will always pass because the calculation becomes purely integer-based. I suggest we keep the current method for variable generation to allow more realistic and diverse in input generation. Btw, this is just the custom choice and have no significant impact to the operation logic, so let's just keep it as it as for now for simplicity.